pkg/mimirtool/commands/rules.go

```diff
@@ -562,61 +597,47 @@ func (r *RuleCommand) syncRules(k *kingpin.ParseContext) error {
 		})
 	}
 
-	err = r.executeChanges(context.Background(), changes)
+	err = r.executeChanges(context.Background(), changes, r.SyncConcurrency)
 	if err != nil {
 		return errors.Wrap(err, "sync operation unsuccessful, unable to complete executing changes")
 	}
 
 	return nil
 }
 
-func (r *RuleCommand) executeChanges(ctx context.Context, changes []rules.NamespaceChange) error {
-	var err error
+func (r *RuleCommand) executeChanges(ctx context.Context, changes []rules.NamespaceChange, concurrencyLimit int) error {
+	// Prepare operations to run, so that we can easily parallelize them.
+	var ops []rules.NamespaceChangeOperation
 	for _, ch := range changes {
-		for _, g := range ch.GroupsCreated {
-			if !r.shouldCheckNamespace(ch.Namespace) {
-				continue
-			}
+		ops = append(ops, ch.ToOperations()...)
+	}
 
-			log.WithFields(log.Fields{
-				"group":     g.Name,
-				"namespace": ch.Namespace,
-			}).Infof("creating group")
-			err = r.cli.CreateRuleGroup(ctx, ch.Namespace, g)
-			if err != nil {
-				return err
-			}
-		}
+	// The execution breaks on first error encountered.
+	err := concurrency.ForEachJob(ctx, len(ops), concurrencyLimit, func(ctx context.Context, idx int) error {
+		op := ops[idx]
 
-		for _, g := range ch.GroupsUpdated {
-			if !r.shouldCheckNamespace(ch.Namespace) {
-				continue
-			}
+		log.WithFields(log.Fields{
+			"group":     op.RuleGroup.Name,
+			"namespace": op.Namespace,
+		}).Infof("synching %s group", op.State.String())
 
-			log.WithFields(log.Fields{
-				"group":     g.New.Name,
-				"namespace": ch.Namespace,
-			}).Infof("updating group")
-			err = r.cli.CreateRuleGroup(ctx, ch.Namespace, g.New)
-			if err != nil {
-				return err
-			}
-		}
+		switch op.State {
+		case rules.Created:
+			return r.cli.CreateRuleGroup(ctx, op.Namespace, op.RuleGroup)
 
-		for _, g := range ch.GroupsDeleted {
-			if !r.shouldCheckNamespace(ch.Namespace) {
-				continue
-			}
+		case rules.Updated:
+			return r.cli.CreateRuleGroup(ctx, op.Namespace, op.RuleGroup)
 
-			log.WithFields(log.Fields{
-				"group":     g.Name,
-				"namespace": ch.Namespace,
-			}).Infof("deleting group")
-			err = r.cli.DeleteRuleGroup(ctx, ch.Namespace, g.Name)
-			if err != nil && !errors.Is(err, client.ErrResourceNotFound) {
-				return err
-			}
+		case rules.Deleted:
+			return r.cli.DeleteRuleGroup(ctx, op.Namespace, op.RuleGroup.Name)
+
+		default:
+			return fmt.Errorf("unsupported operation %q", op.State)
```

Comment:
```suggestion
			return fmt.Errorf("internal error: unexpected operation %q", op.State)
```

What do you think about changing this error message to indicate that it's a bug in Mimirtool? As is, it sounds like a user error, which could be confusing.

Reply:
Makes sense.